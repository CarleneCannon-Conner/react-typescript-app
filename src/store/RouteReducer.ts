import { combineReducers, Reducer } from 'redux'

import { User } from './user/UserTypes'
import { userReducer } from './user/UserReducer'

export interface AppState {
  user: User,
  friendsList: string[],
}

export const rootReducer: Reducer<AppState> = combineReducers<AppState>({
  user: userReducer
} as any)