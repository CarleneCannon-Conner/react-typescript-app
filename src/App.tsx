import React, { ChangeEvent, useEffect, useState } from 'react';
import { Link } from 'react-router-dom'
import { Dispatch } from 'redux'
import { connect, MapDispatchToProps } from 'react-redux'
import {
  saveUserName as saveUserNameAction,
  saveUserMessage as saveUserMessageAction,
} from './store/user/UserActions'
import { User } from './store/user/UserTypes'
import './App.css';
import { AppState } from './store/RouteReducer';

interface AppOwnProps {
  userName: string | undefined
  userType: 'admin' | 'moderator' | 'user' |'guest'
}

interface AppDispatchToProps {
  saveUserName: (user: User) => void
  saveUserMessage: (user: User) => void
}

const App: React.FC<AppDispatchToProps & AppOwnProps> = ({
  userType,
  userName,
  saveUserName,
  saveUserMessage,
}):JSX.Element => {
  const [time, setTime] = useState<Date>(() => new Date(Date.now()))
  const [message, setMessage] = useState<string>('')

  const handleTextChange = (event: ChangeEvent<HTMLInputElement>): void => {
    setMessage(event.target.value)
  }

  useEffect(() => {
    const timer = setInterval(() => {
      setTime(new Date(Date.now()))
    }, 1000)

    if (userName) {
     saveUserName({ userName, userMessage: undefined })
    }

    return () => {
      clearInterval(timer)
    }
  }, [userName, saveUserName])

  useEffect(() => {
    saveUserMessage({ userName: undefined, userMessage: message })
  }, [message, saveUserMessage])


  return (
    <div className="App">
     <p>
       Hi, {userName ? userName : 'Mysterious Entity'}, your user type is {userName ? userType : 'irrelevent because I do not know you'}.
     </p>
     <p>{time.toUTCString()}</p>
     <input
      type='text'
      placeholder='Enteryour message here'
      value={message}
      onChange={handleTextChange}
     />
    <p>
      your message: {message || ''}
    </p>
    <Link to='/userlist'>User List</Link>
    </div>
  );
}

const mapDispatchToProps: MapDispatchToProps<
  AppDispatchToProps,
  AppOwnProps
  > = (dispatch: Dispatch, ownProps: AppOwnProps): AppDispatchToProps => ({
    saveUserName: (user: User) => {
      dispatch(saveUserNameAction(user))
    },
    saveUserMessage: (user: User) => {
      dispatch(saveUserMessageAction(user))
    },
  })

export default connect<
  {}, // AppStateToProps
  AppDispatchToProps,
  AppOwnProps,
  AppState
  >(null, mapDispatchToProps)(App)
