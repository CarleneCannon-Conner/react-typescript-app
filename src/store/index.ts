import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'
import { rootReducer } from './RouteReducer'

export const store = createStore(rootReducer, compose(
  applyMiddleware(thunk),
  ((window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ && (window as any).__REDUX_DEVTOOLS_EXTENSION__()) || compose
))

store.subscribe(() => {
  console.log(`Store State: ${JSON.stringify(store.getState())}`)
})