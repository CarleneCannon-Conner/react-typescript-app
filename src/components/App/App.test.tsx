import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter, Route } from 'react-router-dom';
import App from './App'
import { store } from '../../store'
import { create } from 'react-test-renderer'

const contextProvider = (component: JSX.Element) => 
<Provider store={store}>
      <BrowserRouter>
      <Route
        exact
        path='/'>
          {component}
        </Route>
      </BrowserRouter>
      </Provider>

describe('<App />',() => {
  test('Snapshot test', () => {
    let tree = create(contextProvider(
      <App userType='admin' userName='someone' />
    ))
    expect(tree.toJSON()).toMatchSnapshot()
  })
})