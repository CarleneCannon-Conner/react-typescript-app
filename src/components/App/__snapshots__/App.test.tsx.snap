// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Snapshot test 1`] = `
<div
  className="App"
>
  <p>
    Hi, 
    someone
    , your user type is 
    admin
    .
  </p>
  <p>
    Sat, 07 Aug 2021 16:44:35 GMT
  </p>
  <input
    onChange={[Function]}
    placeholder="Enter your message here"
    type="text"
    value=""
  />
  <p>
    your message: 
    
  </p>
  <a
    href="/userlist"
    onClick={[Function]}
  >
    User List
  </a>
</div>
`;
